### DISCUSSION TITLE
New feature: `SRANDOM` as in bash 5.1+, with upper bound

### DISCUSSION THREAD

**Discussion by McDutchie** [ksh93/ksh#705] [New features]
`SRANDOM` is a secure random number generator. It uses the OS implementation of `arc4random`(3) or, absent that, an integrated OpenBSD-based version that uses the ChaCha20 cipher. An upper bound can be set by assigning its value to SRANDOM; subsequent numbers will be uniformly distributed between 0 and the value of the upper bound minus one, in a way that avoids [modulo bias](https://research.kudelskisecurity.com/2020/07/28/the-definitive-guide-to-modulo-bias-and-how-to-avoid-it/).

I'd like everyone to play with and test this new feature and report back – well, those of you who keep up with the dev branch, that is. It's on the dev branch only and will be in ksh 93u+m/1.1 and up. See 00b296cd42c13143885475468b0cf6340b737230 for all the info.

I am particularly interested in knowing whether this compiles and runs on more obscure systems (Solaris, QNX, AIX, HP-UX, Cygwin) as I am currently unable to test any of these due to broken hardware and lack of access.

**Comment by JohnoKing** [comment]
I can confirm that `SRANDOM` compiles and functions just fine on Cygwin (Windows 11), as well as Haiku, illumos and *BSD.

**Reply by McDutchie** [reply]
Excellent, thank you.

**Reply by JohnoKing** [reply]
While testing a different issue I stumbled across a bug in `SRANDOM` that occurs on Cygwin, but not Linux (filed as #711).
